File: src/c47/logicalOps/rotateBits.c
// SPDX-License-Identifier: GPL-3.0-only
// SPDX-FileCopyrightText: Copyright The WP43 and C47 Authors

/********************************************//**
 * \file rotateBits.c
 ***********************************************/

#include "c47.h"

/********************************************//**
 * \brief regX ==> regL and ASR(regX) ==> regX
 * enables stack lift and refreshes the stack
 *
 * \param[in] unusedButMandatoryParameter uint16_t
 * \return void
 ***********************************************/
void fnAsr(uint16_t numberOfShifts) {
  int32_t i;
  uint64_t sign;

  if(getRegisterDataType(REGISTER_X) == dtShortInteger) {
    if(!saveLastX()) {
      return;
    }

    sign = *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) & shortIntegerSignBit;
    for(i=1; i<=numberOfShifts; i++) {
      if(i == numberOfShifts) {
        if(*(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) & 1) {
          setSystemFlag(FLAG_CARRY);
        }
        else {
          clearSystemFlag(FLAG_CARRY);
        }
      }

      *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) = (*(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) >> 1) | sign;
    }
    *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) &= shortIntegerMask;
  }
  else {
    displayCalcErrorMessage(ERROR_INVALID_DATA_TYPE_FOR_OP, ERR_REGISTER_LINE, REGISTER_X);
    #if (EXTRA_INFO_ON_CALC_ERROR == 1)
      sprintf(errorMessage, "cannot ASR %s", getRegisterDataTypeName(REGISTER_X, true, false));
      moreInfoOnError("In function fnAsr:", errorMessage, NULL, NULL);
    #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
  }
}



/********************************************//**
 * \brief regX ==> regL and SL(regX) ==> regX
 * enables stack lift and refreshes the stack
 *
 * \param[in] unusedButMandatoryParameter uint16_t
 * \return void
 ***********************************************/
void fnSl(uint16_t numberOfShifts) {
  uint64_t w;
  uint32_t base;
  int32_t i;

  if (!getRegisterAsRawShortInt(REGISTER_X, &w, &base) || !saveLastX())
    return;

  for(i=1; i<=numberOfShifts; i++) {
    if(i == numberOfShifts) {
      if(*(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) & shortIntegerSignBit) {
        setSystemFlag(FLAG_CARRY);
      }
      else {
        clearSystemFlag(FLAG_CARRY);
      }
    }
    w <<= 1;
  }
  reallocateRegister(REGISTER_X, dtShortInteger, SHORT_INTEGER_SIZE_IN_BLOCKS, base);
  *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) = w & shortIntegerMask;
}



/********************************************//**
 * \brief regX ==> regL and SR(regX) ==> regX
 * enables stack lift and refreshes the stack
 *
 * \param[in] unusedButMandatoryParameter uint16_t
 * \return void
 ***********************************************/
void fnSr(uint16_t numberOfShifts) {
  uint64_t w;
  uint32_t base;
  int32_t i;

  if (!getRegisterAsRawShortInt(REGISTER_X, &w, &base) || !saveLastX())
    return;

  for(i=1; i<=numberOfShifts; i++) {
    if(i == numberOfShifts) {
      if(*(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) & 1) {
        setSystemFlag(FLAG_CARRY);
      }
      else {
        clearSystemFlag(FLAG_CARRY);
      }
    }

    w >>= 1;
  }
  reallocateRegister(REGISTER_X, dtShortInteger, SHORT_INTEGER_SIZE_IN_BLOCKS, base);
  *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) = w & shortIntegerMask;
}



/********************************************//**
 * \brief regX ==> regL and RL(regX) ==> regX
 * enables stack lift and refreshes the stack
 *
 * \param[in] unusedButMandatoryParameter uint16_t
 * \return void
 ***********************************************/
void fnRl(uint16_t numberOfShifts) {
  int32_t i;
  uint64_t sign;
  uint64_t w;
  uint32_t base;

  if (!getRegisterAsRawShortInt(REGISTER_X, &w, &base) || !saveLastX())
    return;

  for(i=1; i<=numberOfShifts; i++) {
    sign = (w & shortIntegerSignBit) != 0;
    if(i == numberOfShifts) {
      if(sign) {
        setSystemFlag(FLAG_CARRY);
      }
      else {
        clearSystemFlag(FLAG_CARRY);
      }
    }

    w = (w << 1) | sign;
  }
  reallocateRegister(REGISTER_X, dtShortInteger, SHORT_INTEGER_SIZE_IN_BLOCKS, base);
  *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) = w & shortIntegerMask;
}



/********************************************//**
 * \brief regX ==> regL and RR(regX) ==> regX
 * enables stack lift and refreshes the stack
 *
 * \param[in] unusedButMandatoryParameter uint16_t
 * \return void
 ***********************************************/
void fnRr(uint16_t numberOfShifts) {
  int32_t i;
  uint64_t w;
  uint32_t base;

  if (!getRegisterAsRawShortInt(REGISTER_X, &w, &base) || !saveLastX())
    return;

  for(i=1; i<=numberOfShifts; i++) {
    if(i == numberOfShifts) {
      if(w & 1) {
        setSystemFlag(FLAG_CARRY);
      }
      else {
        clearSystemFlag(FLAG_CARRY);
      }
    }

    w = (w >> 1) | ((w & 1) << (shortIntegerWordSize - 1));
  }
  reallocateRegister(REGISTER_X, dtShortInteger, SHORT_INTEGER_SIZE_IN_BLOCKS, base);
  *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) = w & shortIntegerMask;
}



/********************************************//**
 * \brief regX ==> regL and RLC(regX) ==> regX
 * enables stack lift and refreshes the stack
 *
 * \param[in] unusedButMandatoryParameter uint16_t
 * \return void
 ***********************************************/
void fnRlc(uint16_t numberOfShifts) {
  int32_t i;
  uint64_t sign, carry;
  uint64_t w;
  uint32_t base;

  if (!getRegisterAsRawShortInt(REGISTER_X, &w, &base) || !saveLastX())
    return;

  carry = getSystemFlag(FLAG_CARRY);
  for(i=1; i<=numberOfShifts; i++) {
    sign = (w & shortIntegerSignBit) >> (shortIntegerWordSize - 1);
    w = (w << 1) | carry;
    carry = sign;
  }

  if(carry) {
    setSystemFlag(FLAG_CARRY);
  }
  else {
    clearSystemFlag(FLAG_CARRY);
  }
  reallocateRegister(REGISTER_X, dtShortInteger, SHORT_INTEGER_SIZE_IN_BLOCKS, base);
  *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) = w & shortIntegerMask;
}



/********************************************//**
 * \brief regX ==> regL and RRC(regX) ==> regX
 * enables stack lift and refreshes the stack
 *
 * \param[in] unusedButMandatoryParameter uint16_t
 * \return void
 ***********************************************/
void fnRrc(uint16_t numberOfShifts) {
  int32_t i;
  uint64_t lsb, carry;
  uint64_t w;
  uint32_t base;

  if (!getRegisterAsRawShortInt(REGISTER_X, &w, &base) || !saveLastX())
    return;

  carry = getSystemFlag(FLAG_CARRY);
  for(i=1; i<=numberOfShifts; i++) {
    lsb = w & 1;
    w = (w >> 1) | (carry << (shortIntegerWordSize - 1));
    carry = lsb;
  }

  if(carry) {
    setSystemFlag(FLAG_CARRY);
  }
  else {
    clearSystemFlag(FLAG_CARRY);
  }
  reallocateRegister(REGISTER_X, dtShortInteger, SHORT_INTEGER_SIZE_IN_BLOCKS, base);
  *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) = w & shortIntegerMask;
}



/********************************************//**
 * \brief regX ==> regL and LJ(regX) ==> regX
 * enables stack lift and refreshes the stack
 *
 * \param[in] unusedButMandatoryParameter uint16_t
 * \return void
 ***********************************************/
void fnLj(uint16_t unusedButMandatoryParameter) {
  uint32_t count;
  longInteger_t regX;
  uint64_t w;

  if (!getRegisterAsRawShortInt(REGISTER_X, &w, NULL) || !saveLastX())
    return;

  if (w == 0)
    count = shortIntegerWordSize;
  else
    count = __builtin_clzll(w) - (64 - shortIntegerWordSize);

  longIntegerInit(regX);
  uInt32ToLongInteger(count, regX);

  setSystemFlag(FLAG_ASLIFT);
  liftStack();
  convertLongIntegerToLongIntegerRegister(regX, REGISTER_X);
  longIntegerFree(regX);
}



/********************************************//**
 * \brief regX ==> regL and RJ(regX) ==> regX
 * enables stack lift and refreshes the stack
 *
 * \param[in] unusedButMandatoryParameter uint16_t
 * \return void
 ***********************************************/
void fnRj(uint16_t unusedButMandatoryParameter) {
  uint32_t count;
  longInteger_t regX;
  uint64_t w;

  if (!getRegisterAsRawShortInt(REGISTER_X, &w, NULL) || !saveLastX())
    return;
  count = __builtin_ctzll(w | ~shortIntegerMask);

  longIntegerInit(regX);
  uInt32ToLongInteger(count, regX);

  setSystemFlag(FLAG_ASLIFT);
  liftStack();
  convertLongIntegerToLongIntegerRegister(regX, REGISTER_X);
  longIntegerFree(regX);
}



void fnMirror(uint16_t unusedButMandatoryParameter) {
  uint64_t x, r=0;
  uint32_t base;

  if (!getRegisterAsRawShortInt(REGISTER_X, &x, &base) || !saveLastX())
    return;

  if(x != 0) {
    for(uint32_t i=0; i<shortIntegerWordSize; i++) {
      if(x & (1LL << i)) {
        r |= 1LL << (shortIntegerWordSize-i-1);
      }
    }
  }

  reallocateRegister(REGISTER_X, dtShortInteger, SHORT_INTEGER_SIZE_IN_BLOCKS, base);
  *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) = r;
}

/********************************************//**                    //JM vv
 * \brief regX ==> regL and Change Endianness(regX) ==> regX
 * enables stack lift and refreshes the stack
 *
 * \param[in] unusedButMandatoryParameter uint16_t
 * \return void
 ***********************************************/
void fnSwapEndian(uint8_t bitWidth) {
  uint64_t b7,b6,b5,b4,b3,b2,b1,b0;
  uint64_t x;
  uint32_t base;

  if (!getRegisterAsRawShortInt(REGISTER_X, &x, &base) || !saveLastX())
    return;

  //printf("### %d %d",(shortIntegerWordSize & (bitWidth-1)), (shortIntegerWordSize |  (bitWidth-1) ) + 1 );

  b7 = (x & 0xFF00000000000000) >> (64- 8);
  b6 = (x & 0x00FF000000000000) >> (64-16);
  b5 = (x & 0x0000FF0000000000) >> (64-24);
  b4 = (x & 0x000000FF00000000) >> (64-32);
  b3 = (x & 0x00000000FF000000) >> (64-40);
  b2 = (x & 0x0000000000FF0000) >> (64-48);
  b1 = (x & 0x000000000000FF00) >> (64-56);
  b0 = (x & 0x00000000000000FF) >> (64-64);

  if(bitWidth == 8) {
    if(shortIntegerWordSize<16) {fnSetWordSize(16);}
    else
      if( (shortIntegerWordSize & (bitWidth-1)) != 0 ) {fnSetWordSize((shortIntegerWordSize |  (bitWidth-1) ) + 1);}
    switch(shortIntegerWordSize) {
      case 16: x =                                                                               (b0 << 8 ) | b1; break;
      case 24: x =                                                                  (b0 << 16) | (b1 << 8 ) | b2; break;
      case 32: x =                                                     (b0 << 24) | (b1 << 16) | (b2 << 8 ) | b3; break;
      case 40: x =                                        (b0 << 32) | (b1 << 24) | (b2 << 16) | (b3 << 8 ) | b4; break;
      case 48: x =                           (b0 << 40) | (b1 << 32) | (b2 << 24) | (b3 << 16) | (b4 << 8 ) | b5; break;
      case 56: x =              (b0 << 48) | (b1 << 40) | (b2 << 32) | (b3 << 24) | (b4 << 16) | (b5 << 8 ) | b6; break;
      case 64: x = (b0 << 56) | (b1 << 48) | (b2 << 40) | (b3 << 32) | (b4 << 24) | (b5 << 16) | (b6 << 8 ) | b7; break;
      default:break;
    }
  }
  else
  if(bitWidth == 16) {
    if(shortIntegerWordSize<32) {fnSetWordSize(32);}
    else
      if( (shortIntegerWordSize & (bitWidth-1)) != 0 ) {fnSetWordSize((shortIntegerWordSize |  (bitWidth-1) ) + 1);}
    switch(shortIntegerWordSize) {
      case 32: x =                                                     (b1 << 24) | (b0 << 16) | (b3 << 8 ) | b2; break;
      case 48: x =                           (b1 << 40) | (b0 << 32) | (b3 << 24) | (b2 << 16) | (b5 << 8 ) | b4; break;
      case 64: x = (b1 << 56) | (b0 << 48) | (b3 << 40) | (b2 << 32) | (b5 << 24) | (b4 << 16) | (b7 << 8 ) | b6; break;
      default:break;
    }
  }
  reallocateRegister(REGISTER_X, dtShortInteger, SHORT_INTEGER_SIZE_IN_BLOCKS, base);
  *(REGISTER_SHORT_INTEGER_DATA(REGISTER_X)) = x;
}                                                              //JM ^^

File: zig_src/leaf/rotate_bits.zig
const runtime = @import("shortint_runtime.zig");

fn getShiftInput(word: *u64, base: *u32) bool {
    if (!runtime.getRegisterAsRawShortInt(runtime.REGISTER_X, word, base)) {
        runtime.invalidShortIntegerError(runtime.REGISTER_X);
        return false;
    }
    if (!runtime.saveLastX()) {
        return false;
    }
    return true;
}

fn setShiftResult(word: u64, base: u32) void {
    runtime.setRawShortIntegerRegister(runtime.REGISTER_X, base, word & runtime.shortIntegerMask);
}

fn setCarry(enabled: bool) void {
    runtime.forceSystemFlag(runtime.FLAG_CARRY, @as(c_int, @intFromBool(enabled)));
}

fn justifyResultToRegisters(count: u32, base: u32, value: u64) void {
    setShiftResult(value, base);
    runtime.setSystemFlag(runtime.FLAG_ASLIFT);
    runtime.liftStack();
    runtime.setRawShortIntegerRegister(runtime.REGISTER_X, 10, count);
    runtime.convertShortIntegerRegisterToLongIntegerRegister(runtime.REGISTER_X, runtime.REGISTER_X);
}

fn topShift() u6 {
    return @intCast(runtime.shortIntegerWordSize - 1);
}

pub export fn fnAsr(number_of_shifts: u16) callconv(.c) void {
    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    const sign = word & runtime.shortIntegerSignBit;
    var i: u16 = 0;
    while (i < number_of_shifts) : (i += 1) {
        if (i + 1 == number_of_shifts) {
            setCarry((word & 1) != 0);
        }
        word = (word >> 1) | sign;
    }
    setShiftResult(word, base);
}

pub export fn fnSl(number_of_shifts: u16) callconv(.c) void {
    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    var i: u16 = 0;
    while (i < number_of_shifts) : (i += 1) {
        if (i + 1 == number_of_shifts) {
            setCarry((word & runtime.shortIntegerSignBit) != 0);
        }
        word <<= 1;
    }
    setShiftResult(word, base);
}

pub export fn fnSr(number_of_shifts: u16) callconv(.c) void {
    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    var i: u16 = 0;
    while (i < number_of_shifts) : (i += 1) {
        if (i + 1 == number_of_shifts) {
            setCarry((word & 1) != 0);
        }
        word >>= 1;
    }
    setShiftResult(word, base);
}

pub export fn fnRl(number_of_shifts: u16) callconv(.c) void {
    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    var i: u16 = 0;
    while (i < number_of_shifts) : (i += 1) {
        const sign = (word & runtime.shortIntegerSignBit) != 0;
        if (i + 1 == number_of_shifts) {
            setCarry(sign);
        }
        word = (word << 1) | @as(u64, @intFromBool(sign));
    }
    setShiftResult(word, base);
}

pub export fn fnRr(number_of_shifts: u16) callconv(.c) void {
    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    const shift = topShift();
    var i: u16 = 0;
    while (i < number_of_shifts) : (i += 1) {
        const lsb = word & 1;
        if (i + 1 == number_of_shifts) {
            setCarry(lsb != 0);
        }
        word = (word >> 1) | (lsb << shift);
    }
    setShiftResult(word, base);
}

pub export fn fnRlc(number_of_shifts: u16) callconv(.c) void {
    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    var carry: u64 = @intFromBool(runtime.getSystemFlag(@as(i32, @intCast(runtime.FLAG_CARRY))));
    var i: u16 = 0;
    while (i < number_of_shifts) : (i += 1) {
        const sign = @as(u64, @intFromBool((word & runtime.shortIntegerSignBit) != 0));
        word = (word << 1) | carry;
        carry = sign;
    }

    setCarry(carry != 0);
    setShiftResult(word, base);
}

pub export fn fnRrc(number_of_shifts: u16) callconv(.c) void {
    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    const shift = topShift();
    var carry: u64 = @intFromBool(runtime.getSystemFlag(@as(i32, @intCast(runtime.FLAG_CARRY))));
    var i: u16 = 0;
    while (i < number_of_shifts) : (i += 1) {
        const lsb = word & 1;
        word = (word >> 1) | (carry << shift);
        carry = lsb;
    }

    setCarry(carry != 0);
    setShiftResult(word, base);
}

pub export fn fnLj(unused_but_mandatory_parameter: u16) callconv(.c) void {
    _ = unused_but_mandatory_parameter;

    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    var count: u32 = runtime.shortIntegerWordSize;
    if (word != 0) {
        count = @intCast(@clz(word));
        count -= @as(u32, 64 - runtime.shortIntegerWordSize);
        word <<= @as(u6, @intCast(count));
    }

    justifyResultToRegisters(count, base, word);
}

pub export fn fnRj(unused_but_mandatory_parameter: u16) callconv(.c) void {
    _ = unused_but_mandatory_parameter;

    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    var count: u32 = runtime.shortIntegerWordSize;
    if (word != 0) {
        count = @intCast(@ctz(word | ~runtime.shortIntegerMask));
        word >>= @as(u6, @intCast(count));
    }

    justifyResultToRegisters(count, base, word);
}

pub export fn fnMirror(unused_but_mandatory_parameter: u16) callconv(.c) void {
    _ = unused_but_mandatory_parameter;

    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    var result: u64 = 0;
    var index: u8 = 0;
    while (index < runtime.shortIntegerWordSize) : (index += 1) {
        const src_shift: u6 = @intCast(index);
        if ((word & (@as(u64, 1) << src_shift)) != 0) {
            const dst_shift: u6 = @intCast(runtime.shortIntegerWordSize - index - 1);
            result |= @as(u64, 1) << dst_shift;
        }
    }

    setShiftResult(result, base);
}

fn byte(word: u64, shift: u6) u64 {
    return (word >> shift) & 0xff;
}

pub export fn fnSwapEndian(bit_width: u16) callconv(.c) void {
    var word: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&word, &base)) return;

    const b7 = byte(word, 56);
    const b6 = byte(word, 48);
    const b5 = byte(word, 40);
    const b4 = byte(word, 32);
    const b3 = byte(word, 24);
    const b2 = byte(word, 16);
    const b1 = byte(word, 8);
    const b0 = byte(word, 0);

    if (bit_width == 8) {
        if (runtime.shortIntegerWordSize < 16) {
            runtime.fnSetWordSize(16);
        } else if ((runtime.shortIntegerWordSize & @as(u8, @intCast(bit_width - 1))) != 0) {
            runtime.fnSetWordSize((runtime.shortIntegerWordSize | @as(u8, @intCast(bit_width - 1))) + 1);
        }

        switch (runtime.shortIntegerWordSize) {
            16 => word = (b0 << 8) | b1,
            24 => word = (b0 << 16) | (b1 << 8) | b2,
            32 => word = (b0 << 24) | (b1 << 16) | (b2 << 8) | b3,
            40 => word = (b0 << 32) | (b1 << 24) | (b2 << 16) | (b3 << 8) | b4,
            48 => word = (b0 << 40) | (b1 << 32) | (b2 << 24) | (b3 << 16) | (b4 << 8) | b5,
            56 => word = (b0 << 48) | (b1 << 40) | (b2 << 32) | (b3 << 24) | (b4 << 16) | (b5 << 8) | b6,
            64 => word = (b0 << 56) | (b1 << 48) | (b2 << 40) | (b3 << 32) | (b4 << 24) | (b5 << 16) | (b6 << 8) | b7,
            else => {},
        }
    } else if (bit_width == 16) {
        if (runtime.shortIntegerWordSize < 32) {
            runtime.fnSetWordSize(32);
        } else if ((runtime.shortIntegerWordSize & @as(u8, @intCast(bit_width - 1))) != 0) {
            runtime.fnSetWordSize((runtime.shortIntegerWordSize | @as(u8, @intCast(bit_width - 1))) + 1);
        }

        switch (runtime.shortIntegerWordSize) {
            32 => word = (b1 << 24) | (b0 << 16) | (b3 << 8) | b2,
            48 => word = (b1 << 40) | (b0 << 32) | (b3 << 24) | (b2 << 16) | (b5 << 8) | b4,
            64 => word = (b1 << 56) | (b0 << 48) | (b3 << 40) | (b2 << 32) | (b5 << 24) | (b4 << 16) | (b7 << 8) | b6,
            else => {},
        }
    }

    setShiftResult(word, base);
}

pub export fn fnZip(unused_but_mandatory_parameter: u16) callconv(.c) void {
    _ = unused_but_mandatory_parameter;

    var x: u64 = undefined;
    var y: u64 = undefined;
    var base: u32 = undefined;

    if (!runtime.getRegisterAsRawShortInt(runtime.REGISTER_Y, &y, &base)) {
        runtime.invalidShortIntegerError(runtime.REGISTER_Y);
        return;
    }
    if (!getShiftInput(&x, &base)) return;

    var result: u64 = 0;
    var mask: u64 = 1;
    var shift: u8 = 0;
    var index: u8 = 0;
    const half_word = @divFloor(runtime.shortIntegerWordSize, 2);

    while (index < half_word) : (index += 1) {
        result |= (x & mask) << @as(u6, @intCast(shift));
        shift += 1;
        result |= (y & mask) << @as(u6, @intCast(shift));
        mask <<= 1;
    }

    setShiftResult(result, base);
    runtime.adjustResult(runtime.REGISTER_X, true, true, runtime.REGISTER_X, runtime.REGISTER_Y, @as(runtime.calcRegister_t, -1));
}

pub export fn fnUnzip(unused_but_mandatory_parameter: u16) callconv(.c) void {
    _ = unused_but_mandatory_parameter;

    var a: u64 = undefined;
    var base: u32 = undefined;
    if (!getShiftInput(&a, &base)) return;

    runtime.setSystemFlag(runtime.FLAG_ASLIFT);
    runtime.liftStack();

    var x: u64 = 0;
    var y: u64 = 0;
    var mask: u64 = 1;
    var shift: u8 = 0;
    var index: u8 = 0;
    const half_word = @divFloor(runtime.shortIntegerWordSize, 2);

    while (index < half_word) : (index += 1) {
        x |= (a & mask) >> @as(u6, @intCast(shift));
        shift += 1;
        mask <<= 1;
        y |= (a & mask) >> @as(u6, @intCast(shift));
        mask <<= 1;
    }

    runtime.setRawShortIntegerRegister(runtime.REGISTER_Y, base, y & runtime.shortIntegerMask);
    setShiftResult(x, base);
}
